src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -1048,6 +1060,18 @@ private void joinTokenRing(int delay, boolean autoBootstrap, Collection<String>
                 unsafeDisableNode();
                 throw new BootstrappingSafetyException("Bootstrap streaming failed.");
             }
+
+            if(!localSchemaVersion.equals(Schema.instance.getVersion().toString()) || !isSchemaConsistent(localSchemaVersion))
```

Comment:
i wonder if we are being too restrictive by requiring this. as far as i can tell, it's actually okay for the schema to change bootstrap has started as long as the joining node is receiving forwarded writes + the new keyspace/columnfamily. this is already required for correctness in normal bootstraps (without schema changes), so i'm not sure we should be more strict here.

maybe the more "correct" check here is that other nodes see the joining node and are forwarding writes (presumably if the joining node is in their gossip, this is true). i think the current code already tries to do this via sleeping for 30 seconds to let gossip work and checking that gossip has seen the seeds
```
            Gossiper.instance.addLocalApplicationStates(states);
            setMode(Mode.JOINING, "sleeping " + RING_DELAY + " ms for pending range setup", true);
            Uninterruptibles.sleepUninterruptibly(RING_DELAY, TimeUnit.MILLISECONDS);
[...]

        if (!Gossiper.instance.seenAnySeed())
            throw new IllegalStateException("Unable to contact any seeds!");
```
but doesn't appear to guarantee that all writes are forwarded

Reply:
I think your right that this check is more restrictive than we need to be.